enhancements/builds/entitlements.md

```diff
@@ -0,0 +1,295 @@
+---
+title: openshift-builds-improve-use-of-rhel-entitlements
+
+authors:
+  - "@gabemontero"
+  - "@bparees" (provided a lot of input in tracking Jira)
+
+reviewers:
+  - "@bparees"
+  - "@adambkaplan"
+  - "@who else?"
+  
+approvers:
+  - "@bparees"
+  - "@adambkaplan"
+  - "@who else?"
+  
+creation-date: 2020-02-18
+
+last-updated: yyyy-mm-dd
+
+<!-- status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced -->
+status: provisional
+
+see-also:
+  - "/enhancements/builds/volume-secrets.md"  
+  
+---
+
+# A More Automatic Experience For Use of Entitled RHEL Content In OpenShift V4 Builds
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+ > 1. CRI-O either already has or will have an option to specify host directories that should always be mounted 
+>into containers, akin to what the Red Hat Docker Daemon used to provide with OpenShift V3.  If / when that feature
+>is available could have bearing on the implementation options discussed below, or even when to stage in this 
+>enhancement.  Presumably the MCO would need updating to leverage this CRI-O feature.
+> 2. There has been discussion elsewhere around having a more generic notion of "global resources" that any pod can
+>mount.  Would we want this enhancement to tackle that (probably "no")?  Would we gate this enhancement on the 
+>availability of such a feature (probably "no")?  Upstream feature work is implied with such an item.
```

Comment:
Some customers are in favour of namespaced entitlements, they do not like the "cluster entitlement". They even "disable" the BuildConfig altogether.

Reply:
if they disable buildconfigs then they are not the customer this enhancement is targeted at...  this enhancement is targeted at customers who used builds in 3.x and were used to being able to write a Dockerfile that contains a "RUN yum install some-subscription-package" command have it "just work".

The solution for achieving that today in 4.x is a lengthy set of manual steps (I think they were linked in the proposal).

As for namespaced entitlements, well again that wasn't a capability in 3.x, and my first inclination is to say those users may simply have to continue using the manual process for now, but as i noted above we can certainly keep that use case in mind and ensure we don't design something that rules it out as a future extension.

Reply:
I have a change sitting locally that attempts to capture @bparees 's first comment here.

The additional exchanges in this thread around granularity, use of BCs, are more directly touched upon below.  I'll add any needed updates in those sections.

Here is what I have pre-push for the first item:

```
If however such
>an upstream features becomes more likely by the time implementation of this proposal arrives, it might dictate our 
>choosing the CRI-O related path vs. building a path around post-install ConfigMaps/Secrets, as that later path would
>most likely be obviated by the upstream feature.
```

Reply:
We have the automount of secrets from the host into the Pods enabled per default. Some people raised concerns about this default setting. 

If we disable the automount per default and only enable it when a customer/admins want cluster-wide entitled builds, this will make some customers already happy. 

- No MCO interaction to disable the automount which causes the issue with the missing rhsm.conf if a customer/admin does not want cluster-wide entitlements
- The only thing a customer would need to do is create a secret from the downloaded certificates

Not disabling the automount and telling customers to get a rhsm.conf from somewhere and mount the rhsm.conf is cumbersome, what if the config changes, source and target diverge? 

Those steps are all needed because the automount is enabled per default, otherwise, we would have a single understandable step. 

Another annoyance is how the customer can retrieve the certificates "automagically" without going to access.redhat.com.

Reply:
> We have the automount of secrets from the host into the Pods enabled per default. Some people raised concerns about this default setting.

we have this enabled where?  in 4.x?  for what content?

Reply:
```
Starting pod/ip-xxxxxxxxxxx-debug ...
To use host binaries, run `chroot /host`
Pod IP: 10.0.1.58
If you don't see a command prompt, try pressing enter.
sh-4.2# chroot /host
sh-4.4# cat /usr/share/containers/mounts.conf 
/usr/share/rhel/secrets:/run/secrets
sh-4.4# 
```

Per default enabled in RHCOS, thats why I am creating a MachineConfig to disable the automount with an empty `/etc/share/containers/mounts.conf`

In a Pod we have the symlinks to: 
```
bash-4.2$ ls -l /etc/rhsm-host 
lrwxrwxrwx. 1 root root 17 Mar  6  2019 /etc/rhsm-host -> /run/secrets/rhsm

bash-4.2$ ls -l /etc/pki/entitlement-host 
lrwxrwxrwx. 1 root root 32 Mar  6  2019 /etc/pki/entitlement-host -> /run/secrets/etc-pki-entitlement
```

Running dnf/yum  in the container without an rhsm.conf coming from the host (which is missing in RHCOS per default) segfaults. That is why we have to provide an additional rhsm.conf in the enablement. 

RHCOS has the subscription-manager-certs installed but missing the subscription-manger package which carries the rhsm.conf.  (https://bugzilla.redhat.com/show_bug.cgi?id=1783393)

Reply:
Thanks for the detail @zvonkok 

That said, after reading your comments, then [bz 1783393](https://bugzilla.redhat.com/show_bug.cgi?id=1783393), [bz 1785741](https://bugzilla.redhat.com/show_bug.cgi?id=1785741), and the [current tutorial](https://gitlab.cee.redhat.com/dmabe/kvc-mellanox/blob/master/README-entitlements.md) @dustymabe wrote for both day 1 install and day 2 MCO, if there is a definitive path for how we should proceed with disabling automount and/or injecting the right entitlement parts into the MCO, I'm missing it.

Is there consensus on a concise list of steps we should plan for?  If so, can you list them out in a single list?

Did the meeting mentioned in 1785741 happen yet?

thanks again

Reply:
For me the first step would be to disable the default automount so that customers only need the step with the secret. 

If we have established a delivery system for the cluster either via annotations or any other method I am happy to use it. 

The meeting did not happen yet, because this was in idea to patch the subscription-manager, but I do not think that this is the right way.

Reply:
One more follow up question @zvonkok - in your comment https://github.com/openshift/enhancements/pull/214#discussion_r381486918 when you say 

`The only thing a customer would need to do is create a secret from the downloaded certificates`

Are there any requirements for metadata or other settings on the secret from the host so that it will be picked up, assuming automount is disabled, and assuming part of that secret includes the missing rhsm.conf, so we do not get the segfault ?

If it makes sense, could you provide some example yaml or whatever to illustrate what the secret looks like?

Reply:
Or are you perhaps referring to whatever cri-o config option let's you suck in file content from the node by default @zvonkok ?

Maybe that is what `/etc/share/containers/mounts.conf` is ?

I think it is after reading https://github.com/cri-o/cri-o/blob/master/docs/crio.conf.5.md

Reply:
OK, I think I finally have my head around this @zvonkok .  Will have an update shortly.